src/ui/public/courier/data_source/search_source.js

```diff
@@ -51,280 +51,645 @@
+      return this;
     }
 
-    state.index = indexPattern;
-    if (!state.source) {
-      // imply source filtering based on the index pattern, but allow overriding
-      // it by simply setting another value for "source". When index is changed
-      state.source = function () {
-        return indexPattern.getSourceFiltering();
-      };
-      state.source[forIp] = indexPattern;
+    /**
+     * Get the parent of this SearchSource
+     * @return {undefined|searchSource}
+     */
+    getParent(onlyHardLinked) {
+      const self = this;
+      if (self._parent === false) return;
+      if (self._parent) return self._parent;
+      return onlyHardLinked ? undefined : Private(RootSearchSourceProvider).get();
     }
 
-    return this;
-  };
-
-  SearchSource.prototype.extend = function () {
-    return (new SearchSource()).inherits(this);
-  };
-
-  /**
-   * Set a searchSource that this source should inherit from
-   * @param  {SearchSource} searchSource - the parent searchSource
-   * @return {this} - chainable
-   */
-  SearchSource.prototype.inherits = function (parent) {
-    this._parent = parent;
-    return this;
-  };
-
-  /**
-   * Get the parent of this SearchSource
-   * @return {undefined|searchSource}
-   */
-  SearchSource.prototype.getParent = function (onlyHardLinked) {
-    const self = this;
-    if (self._parent === false) return;
-    if (self._parent) return self._parent;
-    return onlyHardLinked ? undefined : Private(RootSearchSourceProvider).get();
-  };
-
-  /**
-   * Temporarily prevent this Search from being fetched... not a fan but it's easy
-   */
-  SearchSource.prototype.disable = function () {
-    this._fetchDisabled = true;
-  };
-
-  /**
-   * Reverse of SourceAbstract#disable(), only need to call this if source was previously disabled
-   */
-  SearchSource.prototype.enable = function () {
-    this._fetchDisabled = false;
-  };
-
-  SearchSource.prototype.onBeginSegmentedFetch = function (initFunction) {
-    const self = this;
-    return new Promise((resolve, reject) => {
-      function addRequest() {
+    /**
+     * Temporarily prevent this Search from being fetched... not a fan but it's easy
+     */
+    disable() {
+      this._fetchDisabled = true;
+    }
+
+    /**
+     * Reverse of SearchSource#disable(), only need to call this if source was previously disabled
+     */
+    enable() {
+      this._fetchDisabled = false;
+    }
+
+    onBeginSegmentedFetch(initFunction) {
+      const self = this;
+      return new Promise((resolve, reject) => {
+        function addRequest() {
+          const defer = Promise.defer();
+          const req = new SegmentedRequest(self, defer, initFunction);
+
+          req.setErrorHandler((request, error) => {
+            reject(error);
+            request.abort();
+          });
+
+          // Return promises created by the completion handler so that
+          // errors will bubble properly
+          return req.getCompletePromise().then(addRequest);
```

Comment:
This appears to be an infinite loop? Not sure if it is or not. If not, is there a way to make this clearer?

Reply:
This is some funky logic for sure.  Take a look at https://github.com/elastic/kibana/issues/16641 specifically this comment - https://github.com/elastic/kibana/issues/16641#issuecomment-370545817.  I'm pretty sure this function is only used by discover and it's how it keeps a request always in the queue so auto refresh works.

We def want to clean this up (or just rewrite all the logic).  Courier is some seriously old code that is pretty tricky to refactor because of all the subtle ways it's used.

For this current PR, I'd prefer to keep it simply the combining of the abstract class and switching to es6 styles (and i think I removed an unused function).  I think it will be easier to isolate bugs if the PRs are relatively focused.  But def a next step will be to do more refactoring/clean up here.